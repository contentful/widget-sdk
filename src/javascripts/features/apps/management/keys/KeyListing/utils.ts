import { getAppDefinitionLoader } from '../../../../apps-core';
import * as util from '../../util';
import moment from 'moment';
import * as ManagementApiClient from '../../ManagementApiClient';

export const fetchKeys = async (orgId, definitionId) => {
  const keys = await getAppDefinitionLoader(orgId).getKeysForAppDefinition(definitionId);

  return Promise.all(keys.map(getFormattedKey));
};

export const getFormattedKey = async (key) => {
  const fingerprint = key.jwk.x5t;
  const hexFingerprint = util.base64ToHex(fingerprint).replace(/(..)/g, '$1:').replace(/:$/, '');
  const mid = hexFingerprint.length / 2 + 1;
  const fingerprintLines = [hexFingerprint.substr(0, mid), hexFingerprint.substr(mid)];

  return {
    fingerprint,
    fingerprintLines,
    createdAt: moment(key.sys.createdAt).format('MMMM DD, YYYY'),
    lastUsedAt: key.sys.lastUsedAt
      ? 'Last used ' + util.formatPastDate(key.sys.lastUsedAt)
      : 'Never used',
    // TODO: batch getting creators
    createdBy: await ManagementApiClient.getCreatorNameOf(key),
  };
};

export const downloadAsFile = (mimeType: string, payload: string, fileName: string) => {
  const temporaryAnchor = document.createElement('a');
  temporaryAnchor.download = fileName;

  const blob = new Blob([payload], {
    type: mimeType,
  });

  temporaryAnchor.href = URL.createObjectURL(blob);

  temporaryAnchor.click();

  URL.revokeObjectURL(temporaryAnchor.href);
};
