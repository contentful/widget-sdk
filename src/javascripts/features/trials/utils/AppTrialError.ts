class ExtendableError extends Error {
  constructor(message?) {
    super(message);
    this.name = this.constructor.name;
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }
}

export class TrialSpaceServerError extends ExtendableError {}
export class ContentImportError extends ExtendableError {}
export class AppInstallationError extends ExtendableError {}
