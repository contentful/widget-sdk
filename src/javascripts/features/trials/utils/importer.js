// eslint-disable-next-line import/no-default-export
export default async function importer() {
  return await import(
    /* webpackChunkName: "experience-provisioning" */ '@contentful/experience-provisioning'
  );
}
