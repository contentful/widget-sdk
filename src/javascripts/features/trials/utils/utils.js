import moment from 'moment';

export const calcTrialDaysLeft = (trialEndDate) => {
  if (!trialEndDate) {
    return;
  }
  return Math.abs(moment().startOf('day').diff(moment(trialEndDate), 'days'));
};
