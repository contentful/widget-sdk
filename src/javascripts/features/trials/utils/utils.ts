import moment from 'moment';
import { getSpaces } from 'services/TokenStore';

export const calcTrialDaysLeft = (trialEndDate?: string | null) => {
  if (!trialEndDate) {
    return -1;
  }
  return Math.abs(moment().startOf('day').diff(moment(trialEndDate), 'days'));
};

export const isSpaceAccessible = async (spaceId?: string) => {
  if (!spaceId) {
    return false;
  }

  const accessibleSpaces = await getSpaces();

  return !!accessibleSpaces.find((space) => space.sys.id === spaceId);
};
