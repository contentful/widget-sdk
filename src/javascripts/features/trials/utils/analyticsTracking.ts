import { track } from 'analytics/Analytics';
import { buildUrlWithUtmParams } from 'utils/utmBuilder';

// these names should be kept up to date with event names defined in transform.js and Validator.js
export const EVENTS = {
  // Click events
  FAIR_USAGE_POLICY: 'fair_use_policy_clicked',
  GET_IN_TOUCH: 'get_in_touch_clicked',
  HELP_LINK: 'help_link_clicked',
  TRIAL_TAG: 'trial_tag_clicked',
  APP_TRIAL_START: 'app_trial_start_clicked',
  // App Trial's performance event
  APP_TRIAL_PERFORMANCE: 'app_trial_start_performance',
} as const;

export const trackEvent = (eventName: typeof EVENTS[keyof typeof EVENTS], data = {}) => {
  track(`trial:${eventName}`, data);
};

export const withInAppHelpUtmParamsSpaceHome = buildUrlWithUtmParams({
  source: 'webapp',
  medium: 'space-trial-home',
  campaign: 'in-app-help',
});

export const withInAppHelpUtmParamsSubscription = buildUrlWithUtmParams({
  source: 'webapp',
  medium: 'subscription-enterprise-trial',
  campaign: 'in-app-help',
});
