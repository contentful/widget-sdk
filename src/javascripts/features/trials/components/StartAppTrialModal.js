import React from 'react';
import PropTypes from 'prop-types';
import { trackEvent, EVENTS } from '../utils/analyticsTracking';
import { Button, Paragraph, Modal } from '@contentful/forma-36-react-components';
import { css } from 'emotion';

const styles = {
  paragraph: css({
    paddingBottom: '10px',
  }),
};

export function StartAppTrialModal({ onConfirm, isShown, onClose }) {
  const handleOnConfirm = () => {
    trackEvent(EVENTS.APP_TRIAL_START)();
    onConfirm();
    onClose();
  };

  return (
    <Modal
      position="center"
      isShown={isShown}
      title={`Compose + Launch free trial`}
      onClose={onClose}>
      {({ title }) => (
        <>
          <Modal.Header title={title} />
          <Modal.Content>
            <Paragraph element="p" className={styles.paragraph}>
              This trial will give you free access to Compose + Launch for 10 days. You can test
              both apps in a new trial space with example content, or you can add some of your own.
              No payment is required, and the trial won&apos;t affect your exisiting spaces or
              content in any way.
            </Paragraph>
          </Modal.Content>
          <Modal.Controls>
            <Button onClick={handleOnConfirm} testId="confirm-button">
              Start your free trial
            </Button>
            <Button onClick={onClose} buttonType="muted">
              Cancel
            </Button>
          </Modal.Controls>
        </>
      )}
    </Modal>
  );
}

StartAppTrialModal.propTypes = {
  isShown: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  onConfirm: PropTypes.func.isRequired,
};
