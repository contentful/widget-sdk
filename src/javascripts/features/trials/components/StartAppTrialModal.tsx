import React from 'react';
import { trackEvent, EVENTS } from '../utils/analyticsTracking';
import { Button, Paragraph, Modal } from '@contentful/forma-36-react-components';
import { css } from 'emotion';

const styles = {
  paragraph: css({
    paddingBottom: '10px',
  }),
};

interface StartAppTrialModalProps {
  isShown: boolean;
  onClose: () => void;
  onConfirm: () => void;
}

export function StartAppTrialModal({ onConfirm, isShown, onClose }: StartAppTrialModalProps) {
  const handleOnConfirm = () => {
    trackEvent(EVENTS.START_APP_TRIAL_MODAL, { elementId: 'confirm_button' });
    onConfirm();
    onClose();
  };

  const handleOnClose = () => {
    trackEvent(EVENTS.START_APP_TRIAL_MODAL, { elementId: 'cancel_button' });
    onClose();
  };

  return (
    <Modal
      position="center"
      isShown={isShown}
      title="Compose + Launch free trial"
      onClose={onClose}>
      {({ title }) => (
        <>
          <Modal.Header title={title} />
          <Modal.Content>
            <Paragraph element="p" className={styles.paragraph}>
              This trial will give you free access to Compose + Launch for 10 days. You’ll get both
              apps in a new trial space where you can use our example content or create your own. No
              payment is required, and the trial wonʼt affect your exisiting spaces or content in
              any way.
            </Paragraph>
          </Modal.Content>
          <Modal.Controls>
            <Button onClick={handleOnConfirm} testId="confirm-button">
              Start your free trial
            </Button>
            <Button onClick={handleOnClose} buttonType="muted" testId="cancel-button">
              Cancel
            </Button>
          </Modal.Controls>
        </>
      )}
    </Modal>
  );
}
