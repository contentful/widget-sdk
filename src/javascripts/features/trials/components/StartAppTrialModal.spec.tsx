import React from 'react';
import { render, waitFor, screen, fireEvent } from '@testing-library/react';
import { track } from 'analytics/Analytics';
import { StartAppTrialModal } from './StartAppTrialModal';
import { EVENTS } from '../utils/analyticsTracking';

const onClose = jest.fn();
const onConfirm = jest.fn();

const build = () => {
  return render(<StartAppTrialModal isShown onClose={onClose} onConfirm={onConfirm} />);
};

jest.mock('analytics/Analytics', () => ({
  track: jest.fn(),
}));

jest.mock('states/Navigator', () => ({
  getCurrentStateName: jest.fn().mockReturnValue('current.state.name'),
}));

describe('StartAppTrialModal', () => {
  it('should render', () => {
    expect(build().baseElement).toMatchSnapshot();
  });

  it('fires off all needed events on confirm btn click', async () => {
    build();

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    await waitFor(() => fireEvent.click(screen.queryByTestId('confirm-button')!));

    expect(track).toHaveBeenCalledWith(`trial:${EVENTS.START_APP_TRIAL_MODAL}`, {
      fromState: 'current.state.name',
      elementId: 'confirm_button',
    });

    expect(onConfirm).toHaveBeenCalled();
    expect(onClose).toHaveBeenCalled();
  });

  it('fires off all needed events on cancel btn click', async () => {
    build();

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    await waitFor(() => fireEvent.click(screen.queryByTestId('cancel-button')!));

    expect(track).toHaveBeenCalledWith(`trial:${EVENTS.START_APP_TRIAL_MODAL}`, {
      fromState: 'current.state.name',
      elementId: 'cancel_button',
    });

    expect(onClose).toHaveBeenCalled();
  });
});
