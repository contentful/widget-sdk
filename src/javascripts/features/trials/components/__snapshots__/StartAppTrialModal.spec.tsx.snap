// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartAppTrialModal should render 1`] = `
.emotion-0 {
  padding-bottom: 10px;
}

<body
  class="Modal__body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="Modal__Modal__portal___1AjJC"
  >
    <div
      class="Modal__Modal__overlay___3gPyC Modal__Modal__overlay--centered___2kDOU Modal__Modal__overlay--after-open___sRqhJ"
    >
      <div
        aria-modal="true"
        class="Modal__Modal__wrap___BjfRz Modal__Modal__wrap--after-open___1XDbx"
        role="dialog"
        style="top: 0px;"
        tabindex="-1"
      >
        <div
          class="Modal__Modal___2S_pS"
          data-test-id="cf-ui-modal"
          style="width: 520px;"
        >
          <div
            class="ModalHeader__ModalHeader___1yD1S"
            data-test-id="cf-ui-modal-header"
          >
            <h1
              class="ModalHeader__ModalHeader__title___3IuOy"
            >
              Compose + Launch free trial
            </h1>
          </div>
          <div
            class="ModalContent__ModalContent___2mf3h"
            data-test-id="cf-ui-modal-content"
          >
            <p
              class="Paragraph__Paragraph___2aO-9 emotion-0"
              data-test-id="cf-ui-paragraph"
            >
              This trial will give you free access to Compose + Launch for 10 days. You’ll get both apps in a new trial space where you can use our example content or create your own. No payment is required, and the trial wonʼt affect your exisiting spaces or content in any way.
            </p>
          </div>
          <div
            class="ModalControls__ModalControls___2bTQx"
            data-test-id="cf-ui-modal-controls"
          >
            <button
              class="Button__Button___1ZfFj Button__Button--primary___JImeO"
              data-test-id="confirm-button"
              type="button"
            >
              <span
                class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
                tabindex="-1"
              >
                <span
                  class="Button__Button__label___3tcOj"
                >
                  Start your free trial
                </span>
              </span>
            </button>
            <button
              class="Button__Button___1ZfFj Button__Button--muted___2Wair"
              data-test-id="cancel-button"
              type="button"
            >
              <span
                class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
                tabindex="-1"
              >
                <span
                  class="Button__Button__label___3tcOj"
                >
                  Cancel
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
